### ISSUE TITLE
_counters not set MongoCheckpointStore

### ISSUE THREAD

**Issue by StevenBlair123** [Eventuous/eventuous#65]
I think the ctor of the MongoCheckpointStore should have something like this in it:

_counters.TryAdd(checkPoint, 0);

Otherwise the dictionary has not been initialised (key not found)

I didn't test this from MongoCheckpointStore, but I built my own one, which was very similar.

**Comment by alexeyzimarev** [comment]
Hmm, have you got a specific case when it fails? The normal sequence is `GetLastChecpoint` (unconditionally), then `StoreCheckpoint`. The `_counters[checkpointId]` entry is initialised in `GetLastCheckpoint`, so I assumed it will always work.

**Comment by StevenBlair123** [comment]
@alexeyzimarev  - I copied your Checkpoint and built my own (I didn't want to use Mongo)

`readonly ConcurrentDictionary<string, int> _counters = new();`

And the first occurance of this being used, as you said is `GetLastCheckpoint`:

 `_counters[checkpointId] = 0;`

I couldn't see any place where the key was initialised, i.e.:

`_counters.Add(checkpointId,0);`

As soon as I ran my code, I got an exception when GetLastCheckpoint attempted to update the counter which did not exist.

**Comment by StevenBlair123** [comment]
@alexeyzimarev  - clearly you are not seeing the same issue, so could it be a version problem? But a later version of this ConcurrentDictionary creates a new entry in the Dictionary rather than throw an Exception?

**Comment by alexeyzimarev** [comment]
`_counters[checkpointId] = 0;` is equivalent to `Add` if the dictionary entry doesn't exist. That's how dictionaries always worked. I am not sure what the issue is in your case. I've never seen it fail.

**Comment by StevenBlair123** [comment]
@alexeyzimarev - yeah, can't argue with that. I have tried replicating the issue and no luck. My checkpoint was slightly different, and I am wondering if I had somehow managed to try and use the key before the GetLastCheckpoint init.
Insider my GetLastCheckpoint, I called another method which loaded my checkpoint from a file.
From there, I must have assumed it was what i thought was a lack of .Add being called.

Every day's a school day...

Thanks for taking the time to look at this.
